packages/upload/test/keyboard-navigation.common.js

```diff
@@ -31,61 +31,108 @@ describe('keyboard navigation', () => {
-    beforeEach(() => {
-      uploadElement.files = [
-        {
-          ...FAKE_FILE,
-          held: true, // Show the start button
-          error: 'Error', // Show the retry button
-        },
-      ];
+    beforeEach(async () => {
+      const fileList = [];
+      for (let i = 0; i < 2; i++) {
+        const file = createFile(1000, 'application/uknown');
+        Object.assign(file, { name: `file-${i}`, held: true, error: 'Error' });
+        fileList.push(file);
+      }
+      uploadElement.files = fileList;
+
+      await nextRender();
+      fileElements = document.querySelectorAll('vaadin-upload-file');
     });
 
     it('should focus on the start button', async () => {
-      const startButton = fileElement.shadowRoot.querySelector('[part=start-button]');
+      const startButton = fileElements[0].shadowRoot.querySelector('[part=start-button]');
 
       await repeatTab(3);
 
-      expect(fileElement.shadowRoot.activeElement).to.equal(startButton);
+      expect(fileElements[0].shadowRoot.activeElement).to.not.equal(null);
+      expect(fileElements[0].shadowRoot.activeElement).to.equal(startButton);
     });
 
     it('should focus on the retry button', async () => {
-      const retryButton = fileElement.shadowRoot.querySelector('[part=retry-button]');
+      const retryButton = fileElements[0].shadowRoot.querySelector('[part=retry-button]');
 
       await repeatTab(4);
 
-      expect(fileElement.shadowRoot.activeElement).to.equal(retryButton);
+      expect(fileElements[0].shadowRoot.activeElement).to.not.equal(null);
+      expect(fileElements[0].shadowRoot.activeElement).to.equal(retryButton);
     });
 
     it('should focus on the clear button', async () => {
-      const removeButton = fileElement.shadowRoot.querySelector('[part=remove-button]');
+      const removeButton = fileElements[0].shadowRoot.querySelector('[part=remove-button]');
 
       await repeatTab(5);
 
-      expect(fileElement.shadowRoot.activeElement).to.equal(removeButton);
+      expect(fileElements[0].shadowRoot.activeElement).to.not.equal(null);
+      expect(fileElements[0].shadowRoot.activeElement).to.equal(removeButton);
+    });
+
+    it('should focus on upload button when last remaining file is removed', async () => {
+      const removeButton = fileElements[0].shadowRoot.querySelector('[part=remove-button]');
+
+      removeButton.click();
+      await nextFrame();
+      removeButton.click();
+      await nextFrame();
+
+      expect(document.activeElement).to.not.equal(null);
+      expect(document.activeElement).to.equal(uploadButton);
+    });
+
+    it('should focus the next file after removing a file', async () => {
+      const removeButton = fileElements[0].shadowRoot.querySelector('[part=remove-button]');
+
+      removeButton.click();
+      await nextFrame();
+
+      const activeElementFileName = document.activeElement.shadowRoot.querySelector('#name').innerText;
+      expect(activeElementFileName).to.equal('file-1');
```

Comment:
It's better to not rely on internal DOM structure but instead use `file` property which is part of public API:

```suggestion
      expect(document.activeElement.file.name).to.equal('file-1');
```